pkg/dataplane/reloadCredentials.go

```diff
@@ -0,0 +1,192 @@
+	defer r.lock.RUnlock()
+	return r.currentValue.GetToken(ctx, options)
+}
+
+func (r *reloadingCredential) start(ctx context.Context, cloud, credentialFile string) {
+	// set up the file watcher, call load() when we see events or on some timer in case no events are delivered
+	fileWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		r.logger.Error(err, errCreateFileWatcher)
+	}
+	// we close the file watcher if adding the file to watch fails.
+	// this will also close the new go routine created to watch the file
+	err = fileWatcher.Add(credentialFile)
+	if err != nil {
+		fileWatcher.Close()
+		r.logger.Error(err, errAddFileToWatcher)
+		return
+	}
+
+	go func() {
+		defer fileWatcher.Close()
+		defer r.ticker.Stop()
+		for {
+			select {
+			case event, ok := <-fileWatcher.Events:
+				if !ok {
+					r.logger.Info("stopping credential reloader since file watcher has no events")
+					return
+				}
+				if event.Op.Has(fsnotify.Write) {
+					if err := r.load(cloud, credentialFile); err != nil {
+						r.logger.Error(err, errLoadCredentials)
+					}
+				}
+			case <-r.ticker.C:
+				if err := r.load(cloud, credentialFile); err != nil {
+					r.logger.Error(err, errLoadCredentials)
+				}
+			case err, ok := <-fileWatcher.Errors:
+				if !ok {
+					r.logger.Info("stopping credential reloader since file watcher has no events")
+					return
+				}
+				r.logger.Error(err, errLoadCredentials)
+			case <-ctx.Done():
+				r.logger.Info("user signaled context cancel, stopping credential reloader")
+				return
+			}
+		}
+	}()
+}
+
+func (r *reloadingCredential) load(cloud, credentialFile string) error {
+	// read the file from the filesystem and update the current value we're holding on to if the certificate we read is newer, making sure to not step on the toes of anyone calling GetToken()
+	byteValue, err := os.ReadFile(credentialFile)
+	if err != nil {
+		return err
+	}
+
+	var nestedCreds swagger.NestedCredentialsObject
+	err = nestedCreds.UnmarshalJSON(byteValue)
+	if err != nil {
+		return err
+	}
+
+	if r.notBefore == "" {
+		r.notBefore = *nestedCreds.NotBefore
+	} else {
+		r.lock.Lock()
+		err, ok := isLoadedCredentialNewer(*nestedCreds.NotBefore, r.notBefore)
+		r.lock.Unlock()
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return nil
+		}
+	}
+
+	var newCertValue *azidentity.ClientCertificateCredential
+	newCertValue, err = getClientCertificateCredential(nestedCreds, cloud)
+	if err != nil {
+		return err
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.currentValue = newCertValue
+	r.notBefore = *nestedCreds.NotBefore
+
+	return nil
+}
+
+func isLoadedCredentialNewer(newCred string, currentCred string) (error, bool) {
+	parsedNewCred, err := time.Parse(time.RFC3339, newCred)
+	if err != nil {
+		return err, false
+	}
+
+	parsedCurrentCred, err := time.Parse(time.RFC3339, currentCred)
+	if err != nil {
+		return err, false
+	}
+
+	return nil, parsedNewCred.After(parsedCurrentCred) || parsedNewCred.Equal(parsedCurrentCred)
```

Comment:
I think you did the `|| Equal()` to handle the base case, but this has the downside that you acquire the write lock and update the current value even when nothing has changed. You could instead do

```
needsUpdate := func(){
    r.lock.Rlock()
    defer r.lock.RUnlock()
    if r.notBefore == "" {
        return true
    } else {
         newer, err := isLoadedCredentialNewer(*nestedCreds.NotBefore, r.notBefore)
        if err != nil {
            r.logger.Error(...)
        }
        return newer
    }
}()
```

and then the `isLoadedCredentialNewer` behavior is simpler.